benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/time_step_mod.f90

```diff
@@ -69,860 +71,808 @@ SUBROUTINE invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, &
 
-        CALL next_sshu_code_set_args( &
-            kernel_next_sshu_code, sshn_u%device_ptr, sshn_t%device_ptr, &
-            sshn_t%grid%tmask_device, sshn_t%grid%area_t_device, &
-            sshn_t%grid%area_u_device)
-        ! Launch the kernel
+        ! Next kernels can be brought before the Copy kernels by switching
+        ! appropriately sshn_t_device <-> ssha_t_device in the set_args
         ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_next_sshu_code, &
             2, C_NULL_PTR, C_LOC(globalsize), C_LOC(localsize), 0, &
             C_NULL_PTR, C_NULL_PTR)
-
-        CALL next_sshv_code_set_args( &
-            kernel_next_sshv_code, sshn_v%device_ptr, sshn_t%device_ptr, &
-            sshn_t%grid%tmask_device, sshn_t%grid%area_t_device, &
-            sshn_t%grid%area_v_device)
-        ! Launch the kernel
         ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_next_sshv_code, &
             2, C_NULL_PTR, C_LOC(globalsize), C_LOC(localsize), 0, &
             C_NULL_PTR, C_NULL_PTR)
 
+        if (.not. no_copy_optimization) then
+            ! The set_args subroutine need to be called every time because the
+            ! field_copy_code kernel is reused with 3 different argument sets.
+            ! Alternatively we could clone the kernel 3 times in allkernels.cl
+            ! with different names.
+            CALL field_copy_code_set_args(kernel_field_copy_code1, &
+                0, SIZE(un%data, 1) - 1, 0, SIZE(un%data, 2) - 1, &
+                un_device, ua_device)
+            ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_field_copy_code1, &
+                2, C_NULL_PTR, C_LOC(globalsize), C_LOC(localsize), 0, &
+                C_NULL_PTR, C_NULL_PTR)
+            CALL field_copy_code_set_args(kernel_field_copy_code2, &
+                0, SIZE(vn%data, 1) - 1, 0, SIZE(vn%data, 2) - 1, &
+                vn_device, va_device)
+            ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_field_copy_code2, &
+                2, C_NULL_PTR, C_LOC(globalsize), C_LOC(localsize), 0, &
+                C_NULL_PTR, C_NULL_PTR)
+            CALL field_copy_code_set_args(kernel_field_copy_code3, &
+                0, SIZE(sshn_t%data, 1) - 1, 0, SIZE(sshn_t%data, 2) - 1, &
+                sshn_t_device, ssha_t_device)
+            ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_field_copy_code3, &
+                2, C_NULL_PTR, C_LOC(globalsize), C_LOC(localsize), 0, &
+                C_NULL_PTR, C_NULL_PTR)
+        else
+            swap = ssha_t%device_ptr
+            ssha_t%device_ptr = sshn_t%device_ptr
+            sshn_t%device_ptr = swap
+
+            swap = ua%device_ptr
+            ua%device_ptr = un%device_ptr
+            un%device_ptr = swap
+
+            swap = va%device_ptr
+            va%device_ptr = vn%device_ptr
+            vn%device_ptr = swap
+        endif
+
         ! Block until all kernels have finished
         ierr = clFinish(cmd_queues(1))
```

Comment:
I see we're using more than one queue earlier. Do we need to call `clFinish` for them too? Either way, please add a comment explaining.

Reply:
Added a comment